src/plugins/events/events.ts

```diff
@@ -69,14 +74,41 @@ export class Events {
     return this.dispatch.length > 0;
   }
 
+  storeMetadata(ctx: Ctx) {
+    this.currentPhase = ctx.phase;
+    this.currentTurn = ctx.turn;
+  }
+
   /**
    * Updates ctx with the triggered events.
    * @param {object} state - The state object { G, ctx }.
    */
   update(state: State) {
     for (let i = 0; i < this.dispatch.length; i++) {
+      const endedTurns = this.currentTurn - this.initialTurn;
+      // This is an arbitrarily large number.
+      const threshold = state.ctx.numPlayers * 10;
+
+      // This protects against potential infinite loops if specific events are called on hooks.
+      // The moment we exceed the defined threshold, we just bail out of all phases.
+      if (endedTurns > threshold) {
+        state = { ...state, ctx: { ...state.ctx, phase: null } };
```

Comment:
On second thoughts, maybe the messed up state might be a better warning until we have something like #962 to allow us to completely cancel the action, but I’d still tend towards just returning `{ ...state }` unless there’s a specific need to set the phase to `null`.

Reply:
I used the same approach that https://github.com/boardgameio/boardgame.io/blob/bfc8a1f2701fd14627c99f38122abc294aa7c16e/src/core/flow.ts#L192. Returning the original state seemed like a good idea too, but I thought it might require more complex changes.

Reply:
I thought that returning the current phase can be confusing when it is not the phase that starts or ends the loop.

Reply:
Ah, great, let’s go with this for consistency then.